### DISCUSSION TITLE
Resize an image to fit within a circular avatar

### DISCUSSION THREAD

**Discussion by hydrajump** [libvips/libvips#2773] [Help wanted]
Hi,

I allow users to upload various image formats (png, jpeg and jpg) and they don't always follow the instructions, so I'm currently manually modifying their images with Figma when necessary.

I specifically ask users to upload a square image (their logo) so that it'll fit in a 400x400 square that I crop as a circle using HTML/CSS.

In an attempt to automate my manual image transformation workflow I discovered libvips and specifically `vipsthumbnail`. Using `vipsthumbnail --size 400x400 /path/to/image` works in most cases except when logos are not square and I still need to manually create a 400x400 white square in Figma and then shrink the user's logo to fit nicely (with some padding) in the center of the square.

It's a simple workflow but tedious.

Any thoughts on how or if libvips can be used to automate this? (Python or Go is what I'm most comfortable with)

Thanks :)

**Comment by jcupitt** [comment]
Hi again,

Yes, vipsthumbnail can resize and crop. Try eg.:

```
vipsthumbnail owl.jpg --size 400x400 --smartcrop attention -o x.jpg
```

With this test image:

![owl](https://user-images.githubusercontent.com/580843/164997714-7243b4b8-150a-4a1e-9072-1fd270285b12.jpg)


I see:

![x](https://user-images.githubusercontent.com/580843/164997719-ae2762dd-b65f-4051-90e9-83375ccc39c4.jpg)

It doesn't always get the crop in the right place :(

**Reply by jcupitt** [reply]
You could put a circular mask on as well, but perhaps that's easiest to do in CSS.

**Reply by jcupitt** [reply]
There's a chapter in the docs about vipsthumbnail, perhaps you saw:

https://www.libvips.org/API/current/Using-vipsthumbnail.md.html

**Reply by hydrajump** [reply]
Sorry for my late! I read about the vipsthumbnail options and gave smart crop a try, but it doesn't work well for my intended use-case which is branding logos. Maybe the solution is to provide the user with a visual representation of what their logo will look like (e.g. similar to Twitter's avatar upload feature) and resize/crop client-side rather than try to use vipsthumbail blindly server-side 🤔 

What do you think?

**Reply by jcupitt** [reply]
You can see libvips smartcrop here:

https://github.com/libvips/libvips/blob/master/libvips/conversion/smartcrop.c#L196-L305

I think I'd just reimplement that, but adjust the weighting a bit.
